try GeometryBasics metadata approaches

visr opened the pull request:
It's nice to use this as a basis for exploring the GeometryBasics approach to metadata, right now for https://github.com/JuliaGeometry/GeometryBasics.jl/issues/49.

Since GeoJSON is a well defined standard, we know exactly what we need. Currently in this PR the switch to GeometryBasics geometries has already been made. To explore the needs for a future metadata in GeometryBasics, I avoided using meta geometries for know, and simply used `Feature` as a composed geometry and named tuple. `FeatureCollection` currently holds a vector of `Feature`s, but should be switched to StructArrays as suggested in https://github.com/JuliaGeometry/GeometryBasics.jl/issues/49.

Supersedes #2

visr committed:
remove GeoInterface integration

visr committed:
Travis CI test only latest julia release for now

visr committed:
move to GeometryBasics

Sov-trotter committed:
Add properties to features in NamedTuple

visr commented on src/GeoJSONTables.jl:
```diff
@@ -18,7 +18,9 @@ function read(source)
     jsonfeatures = get(fc, :features, nothing)
     if get(fc, :type, nothing) == "FeatureCollection" && jsonfeatures isa JSON3.Array
         # TODO fill empty NamedTuple with properties
```
TODO can be removed now

visr commented on test/runtests.jl:
```diff
@@ -39,7 +39,7 @@ featurecollections = [g, multipolygon, realmultipolygon, polyline, point, pointn
         f1, _ = iterate(t)
         @test f1 isa GeoJSONTables.Feature
-        @test isempty(Base.propertynames(f1)) # .== [:cartodb_id, :addr1, :addr2, :park])
+        @test !isempty(Base.propertynames(f1)  .== [:cartodb_id, :addr1, :addr2, :park])
```
This doesn't seem right. You are checking if the comparison is empty.

Could you add some more tests? Now we only know it's not empty, but not that it is working. You could check the keys, values, types etc.

Sov-trotter replied:
Oh yeah sure. I had planned on adding more tests later on, for now I wanted to align to the tests you had added.

visr replied:
We probably redo the testing at some point yeah, but in general I think each new functionality should include at least some minimal tests, others you can assume in many cases that it doesn't work. At least that's how it works for me, most code I write and not test is broken.

Sov-trotter committed:
Fix tests

Sov-trotter committed:
Test out StructArrays functionality

Sov-trotter commented:
The current implemenation is starting to hold well for homogeneous types. 
Wherein `GeoJSONTables.Feature{GeometryBasics.Point{2,Int64},(:a, :b),Tuple{String,Int64}}` comes out of `println(f)` in the `staticschema` method.
    
While for heterogeneous data `Feature{T,(:a, :b),Tuple{String,Int64}}` comes out  or `Feature{T,(:a, :b),Types} where Types where T`
So basically we're getting T/Types instead of concrete types depending on whatever is heterogeneous with a 
`MethodError: no method matching getnamestypes(::Type{Feature{T,(:a, :b),Types} where Types where T})`.

The this type inconsistency occurs while creating the vector of Feature structs -> [Line - 34](https://github.com/visr/GeoJSONTables.jl/blob/meta/src/SA_test.jl#L34)
I tried to overload the `getnamestypes` method with `Any` to get it working, but that didn't help either
@piever Do you think I am missing something here?

visr commented:
I don't fully understand the issue for heterogeneous types. We are using `Feature` with type parameters to have the ability to represent both homogeneous (`Feature{Point{2, Float64}}`) and heterogeneous (`Feature{Any/Union}`) feature collections. The `getnametypes` error seems to be a misspelled function name, it should be `getnamestypes`.

Sov-trotter commented:
Oops! I misspelled `getnamestypes` in the above comment. I have fixed it now.

piever commented on src/SA_test.jl:
```diff
@@ -0,0 +1,39 @@
+
+Feature(x; kwargs...) = Feature(x, values(kwargs))
+
+Base.getproperty(x::Feature, s::Symbol) = s == :geometry ? getfield(x, 1) : getproperty(getfield(x, 2), s) # looking for `s` in `properties`
```
It's incorrect to define this without also overloading `propertynames`.

Sov-trotter replied:
You mean `getnamestypes` should be replaced by `Base.propertynames`?

Sov-trotter replied:
And a separate method for types ?

visr replied:
`Base.propertynames` needs to be implemented for `Feature`, this is separate from the `getnamestypes` function, which is just a little helper function that is only used in the StructArrays README example. See the help of `propertynames` for what it should return.

piever commented:
With these types of custom types, you can't really expect StructArrays to know how to widen things correctly upon collection, or at least, I don't know a good way to do it in practice: how would StructArrays figure out that the first type parameter is linked to the first field? In principle, one could allow the widening mechanism to be customizable, but that is not implemented in StructArrays at the moment.

I think it's much easier to create the `StructArray` from a list of vectors. If you have a `Feature` iterator, you could use `Tables.columntable` (Tables is already a StructArrays dependency, so I guess it's OK to use it) on the iterator to change the storage to "column-based".

So something like

```julia
maketable(iter) = maketable(Tables.columntable(iter)::NamedTuple)

# assuming `geometry` is first in `propertynames(f::Feature)`
maketable(cols::NamedTuple) = maketable(first(cols), Base.tail(cols)) # you could also compute the types here with `Base.tuple_type_tail` and `Base.tuple_type_head`

function maketable(geometry, properties::NamedTuple{names, types}) where {names, types}
    F = Feature{eltype(geometry), names, StructArrays.eltypes(types)}
    return StructArray{F}(; geometry=geometry, properties...)
end
```

Note: this requires overload both `propertynames` and `getproperty` for `Feature`.

Sov-trotter commented:
` In principle, one could allow the widening mechanism to be customizable, but that is not implemented in StructArrays at the moment.`

Not sure if it's correct, but I wonder if it was possible to write a `collect_structarrays` approach for the above method for non-standard data layouts?

visr commented:
Let's keep it as simple as possible, and only focus on constructing it from a list of vectors for now.

Sov-trotter committed:
Construct StructArray from list of vectors

Sov-trotter committed:
refactor tests for giving out geometry with propertynames

Sov-trotter commented:
The new approach works well!  For weird reasons I was getting a `type NamedTuple has no field properties` error. 
Unknowingly I put everything in the old environment(one with the `createinstance`, `staticschema` methods) and the error was gone!
I had it tested a few times, and things work well, so the latest commits include a `StructArrays.staticschema` method. I wonder what must be the reason for it? Maybe it was supposed to be written this way? Will test it a few more times before finally integrating everything?

Sov-trotter commented:
![Screenshot from 2020-07-09 22-28-07](https://user-images.githubusercontent.com/43717431/87071531-f9504a80-c237-11ea-940e-ac66207d18dc.png)

Sov-trotter committed:
Return StructArray of Features; Remove FeatureCollection

Sov-trotter committed:
Remove SA_test.jl

Sov-trotter committed:
Update tests

Sov-trotter committed:
Add tests for getter methods

Sov-trotter committed:
Update tests
Add tests for getter methods


Add maketable and other tests

Sov-trotter committed:
Merge remote-tracking branch 'origin/meta' into meta

Sov-trotter committed:
Test out other geojson data in feature collection

Sov-trotter committed:
Add codecov ; badges for travis, codecov 


Add codecov/badges

Sov-trotter committed:
Refactor tests

Sov-trotter commented:
Should I turn off the codecov comments?

visr commented:
Yes please :)
Can be done with such a file: https://github.com/Deltares/Wflow.jl/blob/master/.codecov.yml

Sov-trotter committed:
Turn off comments; upload codecov report

Sov-trotter committed:
Remove upload command (travis not kicking in)

Sov-trotter commented:
So currently [this](https://github.com/visr/GeoJSONTables.jl/blob/meta/src/GeoJSONTables.jl#L3-L5) is the way we're currently getting in the dependencies. Is there a better/clean way to write this ?

visr commented:
I think that's fine, we can leave it like that.

Sov-trotter commented:
Also the nomenclature might create some confusion [here](https://github.com/visr/GeoJSONTables.jl/blob/meta/src/GeoJSONTables.jl#L52) which returns geometries along with the names of properties present in the `Feature` struct, while this [here](https://github.com/visr/GeoJSONTables.jl/blob/meta/src/GeoJSONTables.jl#L58) will return only properties/metadata. 

So it will be weird to get geometry and metadata names out of `Base.propertynames()` and only metadata out of `properties()`

Sov-trotter committed:
Add support for missing types

Sov-trotter committed:
Update Tests

Sov-trotter committed:
Minor Refactoring

Sov-trotter committed:
Add tests for helper methods

Sov-trotter committed:
Test left out GB cases

Sov-trotter committed:
Update Tests


Minor Refactoring


Add tests for helper methods


Test left out GB cases


Cover corner cases

Sov-trotter committed:
Merge remote-tracking branch 'origin/meta' into meta

Sov-trotter committed:
Check for FeatureCollection first

Sov-trotter committed:
Give out JSON3.object too

Sov-trotter committed:
Support for non-tabular Features

Sov-trotter committed:
Add tests

Sov-trotter committed:
Test a borderline case

Sov-trotter committed:
Print a FeatureCollection instead of StructArray

Sov-trotter committed:
Revert "Add test for fc"

This reverts commit 7b7cce3e0b6c3ffbdbef31e37d085966e64f1119.

Sov-trotter committed:
Add test for fc

visr commented on src/GeoJSONTables.jl:
```diff
@@ -101,10 +101,13 @@ function miss(x)
     end
 end
 
+Base.show(io::IO, fc::StructArray) = println(io, "FeatureCollection with $(length(fc)) Features")
```
This overwrites the show method that is defined in StructArrays. We shouldn't do that.

Sov-trotter replied:
Oh. I only found one here https://github.com/JuliaArrays/StructArrays.jl/blob/4483b57939f82828be51bc1e87c4a2de5f49ccf5/src/lazy.jl#L19 for LazyRow not StructArray so went ahead with it?

Sov-trotter replied:
AH, found it https://github.com/JuliaArrays/StructArrays.jl/blob/4483b57939f82828be51bc1e87c4a2de5f49ccf5/src/structarray.jl#L223

visr replied:
Hmm yeah you maybe right that it is not defined in StructArrays. But still it's not good practice to overload base functions on types you don't own.

Sov-trotter committed:
Revert Base piracy

Sov-trotter committed:
Update Readme 1

Sov-trotter committed:
Update Readme 

Update Readme

Sov-trotter committed:
Merge remote-tracking branch 'origin/meta' into meta

visr commented on .travis.yml:
```diff
@@ -1,7 +1,6 @@
-  - 1.3
-  - nightly
+  - 1
+after_success:
```
can be removed

Sov-trotter replied:
I am not sure what you're referring to here?

visr replied:
Just the line. It seems like `after_success:` can be removed since, it's an empty block?

Sov-trotter replied:
The block isn't empty I guess.` codecov:true` is inside it?

visr replied:
That's not insided, because it is not indented.

Sov-trotter replied:
Ah yeah!

visr commented on .travis.yml:
```diff
@@ -1,7 +1,6 @@
+  - 1
+after_success:
+codecov: true
+# - julia -e 'cd(Pkg.dir("GeoJSONTables")); Pkg.add("Coverage"); using Coverage; Codecov.submit(Codecov.process_folder())'
```
can be removed

visr commented on README.md:
```diff
@@ -1,33 +1,61 @@
+This package is unregistered and in development, so expect changes. 
+For now it supports reading, of the following 
+* GeoJSON FeatureCollections
+* [GeometryBasics](https://github.com/JuliaGeometry/GeometryBasics.jl) geometries as Features 
```
It's best to not mix up geometries and features since in GeoJSON they are separate well defined things.

visr commented on README.md:
```diff
@@ -1,33 +1,61 @@
 
 julia> jsonbytes = read("path/to/a.geojson");
 
+#= 
+   A true flag can be passed in the read function 
+   to read the jsonbytes as a JSON3 dict 
+   the default, false, will read it into a StructArray 
+=#
```
I'm afraid that having a boolean flag like this is not really clear to users, and that it is better to offer it in a separate function altogether. Ah but if it just returns a JSON3 Object, people can just use the JSON3 function right? We could point to it in the docs for if people need lower level access though.

Sov-trotter replied:
So it's better to remove this entirely?

visr commented on README.md:
```diff
@@ -1,33 +1,61 @@
 julia> df = DataFrame(fc)
+
+# GeometryBasics geometries can be passed along with metadata, into a Feature
+julia> f = GeoJSONTables.geo2feat(Point(1.0, 2.0), city = "Mumbai", rainfall = 1010)
```
It looks like this could also be a constructor for `Feature`: `GeoJSONTables.Feature(Point(1.0, 2.0), city = "Mumbai", rainfall = 1010))` Does this already work?

Sov-trotter replied:
Yeah. https://github.com/visr/GeoJSONTables.jl/blob/meta/src/GeoJSONTables.jl#L118-L126

Sov-trotter replied:
So `geo2feat()` will handle it, even the `missing` cases nicely.

visr replied:
I think the link is outdated now, not sure what lines you were referring to. It's good to press y in github before copying the link, that will put the commit hash in there such that the lines will stay the ones you mean.

But I don't understand your last comment. So `geo2feat()` does things that `Feature()` doesn't related to `missing`? What is that exactly? And can we put that behavior in the `Feature` constructor as well?

Sov-trotter replied:
Okay. Thanks!
So the link was referring to `geo2feat` methods https://github.com/visr/GeoJSONTables.jl/blob/a8fc7bbfa61045dd1103ff9c5fd69a712b86a2fe/src/GeoJSONTables.jl#L105-L113

So what we want is, like you said earlier that even something like a `Point(1, 2)` is a valid geojson format and we should accept it as well. And I also wanted to be consistent with the currently existing `GeoJSON.jl` package. [This](https://github.com/visr/GeoJSONTables.jl/issues/5#issuecomment-659621092 ) comment includes the details. 

We can also use the existing `Feature` constructor as well. But I was concerned that it might run into conflicts with the actual usage of Feature [here](https://github.com/visr/GeoJSONTables.jl/blob/a8fc7bbfa61045dd1103ff9c5fd69a712b86a2fe/src/GeoJSONTables.jl#L34-L41) for which the [constructor](https://github.com/visr/GeoJSONTables.jl/blob/a8fc7bbfa61045dd1103ff9c5fd69a712b86a2fe/src/GeoJSONTables.jl#L12) already exists.

visr replied:
> for which the constructor already exists

Luckily we have multiple dispatch

> Point(1, 2) is a valid geojson format and we should accept it as well

True, but that should just be a point and not a Feature. I don't see what `geo2feat` does in general, other than acting as a `Feature` constructor.

Sov-trotter replied:
Ah sure. I'll try it out then.

visr commented on src/GeoJSONTables.jl:
```diff
@@ -1,52 +1,73 @@
+    if get(fc, :type, nothing) == "FeatureCollection" && jsonfeatures isa JSON3.Array
+        for f in jsonfeatures 
+            prop = f.properties
+            if !=(get(f, :properties, nothing), nothing)
```
Nit: I prefer the inline `a != b`, think it is a bit clearer, see also below.

Sov-trotter replied:
Yeah. So vscode suggested me this. xD 
>New types should generally not implement this, and rely on the fallback definition !=(x,y) = !(x==y) instead.

visr commented on src/GeoJSONTables.jl:
```diff
@@ -1,52 +1,73 @@
+reads raw jsonbytes into a StructArray table
+"""
+function readtable(source)
+    features = Feature[]
```
It would be more efficient to directly build vectors, one for the geometries, and one for each field, rather than a vector of features that will still end up as the above in memory later on when you create the structarray.

Sov-trotter replied:
But in that case we won't be able to construct the `Feature`? You're referring to this https://github.com/JuliaGeometry/GeometryBasics.jl/issues/49#issuecomment-651203611 right?

visr replied:
Yeah indeed that's what I mean. But Pietro has shown us how to construct the StructArray from vectors with Feature rows right, like here: https://github.com/JuliaGeometry/GeometryBasics.jl/issues/49#issuecomment-654120668?

Sov-trotter replied:
So going by this we will get properties as a StructArray too, hence nested, are we okay with it?

Sov-trotter replied:
One thing we can do here is, test out both the implementations for performance and settle for the performant one?

visr replied:
I'd prefer the non nested approach if we can get that to work.

Sov-trotter replied:
Yeah, the [current](https://github.com/visr/GeoJSONTables.jl/blob/meta/src/Struct_Arrays.jl) approach is non-nested indeed. [This](https://github.com/JuliaArrays/StructArrays.jl#advanced-structures-with-non-standard-data-layout) is what @piever suggested with some [tweaking](https://github.com/visr/GeoJSONTables.jl/pull/3#issuecomment-656099490)

visr commented on src/GeoJSONTables.jl:
```diff
@@ -55,33 +76,57 @@ Returns missing for null/nothing or not present, to work nicely with
-    return Feature(val), state
+function Base.getproperty(f::Feature, s::Symbol)
+    if s == :geometry
+        val = getfield(f, 1)
```
Could you use the above defined `geometry` function here?

visr commented on src/GeoJSONTables.jl:
```diff
@@ -55,33 +76,57 @@ Returns missing for null/nothing or not present, to work nicely with
+        val = getfield(f, 1)
+        miss(val)
+    else
+        val = getproperty(getfield(f, 2), s)
```
Could you use the above defined `properties` function here?

Sov-trotter committed:
Minor refactoring

Sov-trotter committed:
Remove read() flag; update readme for JSON3 interface

visr commented on src/GeoJSONTables.jl:
```diff
@@ -35,7 +35,7 @@ function readtable(source)
         for f in jsonfeatures 
             prop = f.properties
-            if !=(get(f, :properties, nothing), nothing)
+            if get(f, :properties, nothing) != nothing
```
For `nothing` it can be faster to do `=== / !==` versus `== / !=`. You can also use `isnothing`. Note in the lines below there are still a few `!=(` cases as well.

Sov-trotter committed:
Update travis

Sov-trotter committed:
!== over !=

Sov-trotter committed:
Get rid of geo2feat; integrate functionality in Feature

Sov-trotter committed:
Fix interiors bug

visr commented:
Closing as we've decided to keep the wrapped JSON3 Objects as a kind of lazy geometry type, and implement GeoInterface on them, such that we can easily convert to GeometryBasics or other more useful geometry types.
